utils/indexed_datasets.py

```diff
@@ -1,50 +1,49 @@
-import os.path
-import pickle
-from copy import deepcopy
+import pathlib
+import multiprocessing
+from collections import deque
 
+import h5py
+import torch
 import numpy as np
 
 
 class IndexedDataset:
     def __init__(self, path, prefix, num_cache=0):
         super().__init__()
-        self.path = path
-        self.data_file = None
-        self.data_offsets = np.load(os.path.join(path, f'{prefix}.idx'))
-        self.data_file = open(os.path.join(path, f'{prefix}.data'), 'rb', buffering=-1)
-        self.cache = []
+        self.path = pathlib.Path(path)
+        self.dset = h5py.File(self.path / f'{prefix}.hdf5', 'r')
+        self.cache = deque(maxlen=num_cache)
         self.num_cache = num_cache
 
     def check_index(self, i):
-        if i < 0 or i >= len(self.data_offsets) - 1:
+        if i < 0 or i >= len(self.dset):
             raise IndexError('index out of range')
 
     def __del__(self):
-        if self.data_file:
-            self.data_file.close()
+        if self.dset:
+            del self.dset
 
     def __getitem__(self, i):
         self.check_index(i)
         if self.num_cache > 0:
             for c in self.cache:
                 if c[0] == i:
                     return c[1]
-        self.data_file.seek(self.data_offsets[i])
-        b = self.data_file.read(self.data_offsets[i + 1] - self.data_offsets[i])
-        item = pickle.loads(b)
+        item = {k: v[()] if v.shape == () else torch.from_numpy(v[()]) for k, v in self.dset[str(i)].items()}
         if self.num_cache > 0:
-            self.cache = [(i, deepcopy(item))] + self.cache[:-1]
+            self.cache.appendleft((i, item))
         return item
 
     def __len__(self):
-        return len(self.data_offsets) - 1
+        return len(self.dset)
 
 class IndexedDatasetBuilder:
     def __init__(self, path, prefix, allowed_attr=None):
-        self.path = path
+        self.path = pathlib.Path(path)
         self.prefix = prefix
-        self.out_file = open(os.path.join(path, f'{prefix}.data'), 'wb')
-        self.byte_offsets = [0]
+        self.dset = h5py.File(self.path / f'{prefix}.hdf5', 'w')
```

Comment:
Consider using `.h5` which is more common with HDF5 files or `.data` that is the same as before?

Reply:
Use `.data` may confuse the use of previously binarized dataset, I think we stick to `.h5`